src/opcode.rs

```diff
@@ -1369,6 +1369,39 @@ opcode!(
     }
 );
 
+// === 5.18 ===
+
+opcode!(
+    /// Send a message (with data) to a target ring.
+    pub struct MsgRingData {
+        ring_fd: { impl sealed::UseFd },
+        result: { i32 },
+        user_data: { u64 },
+        user_flags: { Option<u32> },
```

Comment:
Note for reviewers: the flags-passing logic will only land in Linux 6.3, but I already added it in the signature here in order to skip a semver breaking change later.

Reply:
If it is an optional flag, I think it should go under `;;` ?

like

```rust
    pub struct MsgRingData {
        ring_fd: { impl sealed::UseFd },
        result: { i32 },
        user_data: { u64 },
        ;;
        user_flags: u32 = 0,
        opcode_flags: u32 = 0
    }
```

`Option` seems unnecessary, because if `None` means it is 0.

Reply:
It is slightly more complex than that. Looking at the [referenced kernel patch](https://lore.kernel.org/all/20230103160507.617416-1-leitao@debian.org/), there is an explicit new flag `IORING_MSG_RING_FLAGS_PASS` that needs to be set on the SQE.

So effectively `Some(0)` and `None` have different behaviors, because the former also sets a flag (which only newer kernels will understand) while the latter does not (thus it works also on current kernels).